src/jit/morph.cpp

```diff
@@ -7498,270 +7475,529 @@ GenTree* Compiler::fgMorphPotentialTailCall(GenTreeCall* call)
+        // materialize as embedded stmts in right execution order.
+        assert(thisPtr != nullptr);
+        call->gtCallArgs = gtPrependNewCallArg(thisPtr, call->gtCallArgs);
+        call->fgArgInfo  = nullptr;
     }
 
-    // Add the extra VSD parameter if needed
-    if (call->IsVirtualStub())
+    // We may need to pass the target address, for instance for calli.
+    if (help.flags & CORINFO_TAILCALL_STORE_TARGET)
     {
-        GenTree* stubAddrArg = fgGetStubAddrArg(call);
-
-        // We don't need this arg to be in the normal stub register, so
-        // clear out the register assignment.
-        assert(stubAddrArg->gtRegNum == virtualStubParamInfo->GetReg());
-        stubAddrArg->gtRegNum = REG_NA;
+        JITDUMP("VM requested target, so adding target\n");
+        GenTree* target;
+        if (call->gtCallType == CT_INDIRECT)
+        {
+            noway_assert(call->gtCallAddr != nullptr);
 
-        // And push the stub address onto the list of arguments
-        call->gtCallArgs = gtPrependNewCallArg(stubAddrArg, call->gtCallArgs);
-    }
-    else if (call->IsVirtualVtable())
-    {
-        noway_assert(thisPtr != NULL);
+            // The target is the last argument and thus evaluation order is
+            // preserved.
+            target           = call->gtCallAddr;
+            call->gtCallAddr = nullptr;
+        }
+        else
+        {
+            noway_assert(call->gtCallType == CT_USER_FUNC && call->gtCallMethHnd != nullptr);
+            target = fgGetDirectCallTargetAddress(call);
+        }
 
-        GenTree* add  = gtNewOperNode(GT_ADD, TYP_I_IMPL, thisPtr, gtNewIconNode(VPTR_OFFS, TYP_I_IMPL));
-        GenTree* vtbl = gtNewOperNode(GT_IND, TYP_I_IMPL, add);
-        vtbl->gtFlags |= GTF_EXCEPT;
+        // Insert target as last arg
+        GenTreeCall::Use** newArgSlot = &call->gtCallArgs;
+        while (*newArgSlot != nullptr)
+        {
+            newArgSlot = &(*newArgSlot)->NextRef();
+        }
 
-        unsigned vtabOffsOfIndirection;
-        unsigned vtabOffsAfterIndirection;
-        bool     isRelative;
-        info.compCompHnd->getMethodVTableOffset(call->gtCallMethHnd, &vtabOffsOfIndirection, &vtabOffsAfterIndirection,
-                                                &isRelative);
+        *newArgSlot = new (this, CMK_ASTNode) GenTreeCall::Use(target, nullptr);
 
-        /* Get the appropriate vtable chunk */
+        call->fgArgInfo  = nullptr;
+    }
 
-        if (vtabOffsOfIndirection != CORINFO_VIRTUALCALL_NO_CHUNK)
-        {
-            add = gtNewOperNode(GT_ADD, TYP_I_IMPL, vtbl, gtNewIconNode(vtabOffsOfIndirection, TYP_I_IMPL));
+    // This is now a direct call to the store args stub and not a tailcall.
+    call->gtCallType    = CT_USER_FUNC;
+    call->gtCallMethHnd = help.hStoreArgs;
+    call->gtFlags &= ~GTF_CALL_VIRT_KIND_MASK;
+    call->gtCallMoreFlags &= ~(GTF_CALL_M_TAILCALL | GTF_CALL_M_DELEGATE_INV | GTF_CALL_M_SECURE_DELEGATE_INV);
```

Comment:
I had a question about this at some point which seems to have disappeared. Would it make sense to unconditionally set this to 0 instead? I am worried that I am missing resetting some flags. However, in the other case I am worried that I am missing preserving some flags...